🚀 [deploy] #20 develop 브랜치 자동화 파이프라인 구축

yeeun0702 opened the pull request:
## 🍞 Issue

Closes #20 

CI/CD 자동화 및 Docker 기반 EC2 배포 파이프라인 구축


## 🥐 Todo

### ✅ CI

- [ ]  `.github/workflows/ci.yml` 작성 및 적용
    - Java 17 설치 (Temurin)
    - `.env` 파일 로드 (GitHub Secrets 활용)
- [ ] `main` , `develop` 브랜치 push/PR 발생 시 CI 수행
- [ ] Gradle 캐시(Cache) 적용: 의존성 다운로드 시간 단축

### ✅ CD

- [ ]  멀티모듈 중 `bbangzip-api`만 빌드되도록 `bootJar` 설정
- [ ]  멀티스테이지 Dockerfile 작성 (build → run)
- [ ]  `develop` 브랜치 push 시 CD 워크플로 실행 (`dev-cd.yml`)
- [ ]  Docker 이미지 빌드 → DockerHub 푸시
- [ ]  EC2 서버에 SSH 접속 → 기존 컨테이너 제거 및 재배포
- [ ]  nginx Docker화 및 `docker-compose` 통합 배포 예정

---

## 🧇 Details

### ✅ CI 구성 (`ci.yml`)

- `develop` 브랜치 기준으로 CI 실행
- `.env` 파일은 GitHub Secrets (`ENV_TEST_FILE`)로 관리
- Gradle 캐시를 적용해 빌드 시간 최적화

```yaml
on:
  push:
    branches: [develop]
  pull_request:
    branches: [develop]
```

### ✅ CD 구성 (`dev-cd.yml`)

- `develop` 브랜치에 푸시 시 자동 배포
- `bbangzip-api`만 `bootJar`로 빌드됨
- Dockerfile은 멀티스테이지 방식 사용 → 빌드용 + 실행용 이미지 분리
- 실행 시 `application-dev.yml`을 로드하도록 설정됨
    
    `CMD ["java", "-jar", "bbangzip.jar", "--spring.config.name=application-dev"]`
    
- EC2에는 nginx 직접 설치됨
    - 80 → 8080 리버스 프록시
    - 추후 Docker로 이관 예정

## 🧇 이후 작업 계획

- `main` 브랜치 기반 GitHub Release 배포 워크플로(prod-release.yml) 구성 예정
- 무중단 배포 + 블루그린 배포 전략 적용
    - Nginx와 `docker-compose-prod.yml`을 활용해 트래픽 전환 방식 설계
- 기존 서비스 중단 없이 새로운 컨테이너로 롤아웃 후, 헬스체크 통과 시 포트 스왑
- 운영 인프라 분리를 위한 VPC 및 서브넷 구성 예정

## 🍩 Reviewer Notes

CI/CD 자동화 및 배포 파이프라인 구성 PR입니다.
구성 흐름이나 설정 상 이상한 부분 있으면 편하게 코멘트 부탁드립니다!  🙌

yeeun0702 committed:
🚀 [deploy] #20 Dockerfile 및 docker-compose-dev.yml 추가

yeeun0702 committed:
🚀 [deploy] #20 application-dev.yml 추가

yeeun0702 committed:
🚀 [deploy] #20 GitHub Actions CI workflow

yeeun0702 committed:
🚀 [deploy] #20 GitHub Actions dev CD workflow

yeeun0702 committed:
🚀 [deploy] #20 GitHub Actions에서 docker-compose 파일 경로 인식 문제 수정

yeeun0702 committed:
🚀 [deploy] #20 GitHub Actions에서 docker-compose 파일 경로 인식 문제 수정

choyeongju commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,48 @@
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      # 2. .env 환경변수 파일 생성 (GitHub Secrets에 저장된 문자열을 파일로 저장)
+      - name: Load .env
+        run: echo "${{ secrets.ENV_TEST_FILE }}" > .env
```
p5. 나중에 develop, main 브랜치에 따라. env 파일 각각 적용하는 분기처리가 필요하겠군요!

choyeongju commented on .github/workflows/dev-cd.yml:
```diff
@@ -0,0 +1,98 @@
+name: CD – Deploy on develop
```
p5. main.yml도 추후 필요하겠군요.. 으앙

choyeongju commented on .github/workflows/dev-cd.yml:
```diff
@@ -0,0 +1,98 @@
+          script: |
+            cd /home/${{ secrets.EC2_USER }}/apps/${{ secrets.PROJECT_NAME }}
+
+            # 🔧 8080 포트 사용 중인 프로세스가 있으면 강제 종료
```
8080으로 찾는게 아닌 컨테이너/서비스 이름 기준으로 찾아서 종료하는 방향은 어떨까유..?!

yeeun0702 replied:
기존에 8080 포트 기준으로 컨테이너를 종료했던 건,  
수동 배포로 띄운 컨테이너가 있어서 포트 충돌을 해결하려구 했던 거 였어요 !

지금은 해당 수동 컨테이너를 정리하고, 컨테이너 이름 기준으로 종료하도록 변경해서 CI/CD로만 관리되도록 구조를 정리했습니당 !

choyeongju commented on Dockerfile:
```diff
@@ -0,0 +1,23 @@
+COPY . .
+
+# bbangzip-api만 빌드
+RUN ./gradlew :bbangzip-api:bootJar
```
`--no-daemon --build-cache` 옵션 추가 어떨까요!!

choyeongju commented on Dockerfile:
```diff
@@ -0,0 +1,23 @@
+
+WORKDIR /app
+
+COPY . .
```
필요한 부분만 copy 도 괜찮을지두..?!

yeeun0702 replied:
Dockerfile 쪽은 전체 복사 대신

- gradle 설정 파일만 먼저 COPY하고
- 실제 필요한 모듈(`bbangzip-api`, `domain`, `external`, `common`, `auth`)만 골라서 COPY하도록 수정했어요!

빌드 속도랑 캐시 활용 면에서 훨씬 좋아진 것 같습니당

choyeongju commented on docker-compose-dev.yml:
```diff
@@ -0,0 +1,17 @@
+    ports:
+      - "8080:8080"
+    env_file:
+      - .env
```
상대경로가 더 안전할 것 같슴도리!

choyeongju reviewed (commented):
너무 수고하셨슴니다.. ㅠㅠ 최고 최고 예은이 최고

yeeun0702 committed:
🚀 [deploy] #20 --no-daemon --build-cache 옵션 추가

yeeun0702 committed:
🚀 [deploy] #20 필요한 모듈만 COPY

yeeun0702 committed:
🚑 [hotfix] #20 bbangzip-api/build.gradle 충돌 해결

Merge branch 'develop' of https://github.com/Team-BBANGZIP/BBANGZIP-SERVER into deploy/20

# Conflicts:
#	bbangzip-api/build.gradle

yeeun0702 committed:
🚀 [deploy] #20 bbangzip-api bootJar 파일명 고정
